crates/table/src/table.rs

```diff
@@ -234,20 +250,242 @@ impl Table {
+                // SAFETY:
+                // - `elem_ty` appears as a column in th row type.
+                // - `AlgebraicValue` is compatible with all types.
+                let val = unsafe { AlgebraicValue::unchecked_read_column(row_ref, elem_ty) };
+                val.is_numeric_zero()
+            })
+            .map(|seq| (seq.col_pos, seq.sequence_id))
+            .unzip()
+    }
+
+    /// Writes `seq_val` to the column at `col_id` in the row identified by `ptr`.
+    ///
+    /// Truncates the `seq_val` to fit the type of the column.
+    ///
+    /// # Safety
+    ///
+    /// - `self.is_row_present(row)` must hold.
+    /// - `col_id` must be a valid column, with a primiive type, of the row type.
+    pub unsafe fn write_gen_val_to_col(&mut self, col_id: ColId, ptr: RowPointer, seq_val: i128) {
+        let row_ty = self.inner.row_layout.product();
+        let elem_ty = unsafe { row_ty.elements.get_unchecked(col_id.idx()) };
+        let AlgebraicTypeLayout::Primitive(col_typ) = elem_ty.ty else {
+            // SAFETY: Columns with sequences must be primitive types.
+            unsafe { unreachable_unchecked() }
+        };
+
+        let fixed_row_size = self.inner.row_layout.size();
+        let fixed_buf = self.inner.pages[ptr.page_index()].get_fixed_row_data_mut(ptr.page_offset(), fixed_row_size);
+
+        fn write<const N: usize>(dst: &mut [u8], offset: u16, bytes: [u8; N]) {
+            let offset = offset as usize;
+            dst[offset..offset + N].copy_from_slice(&bytes);
+        }
+
+        match col_typ {
+            PrimitiveType::I8 => write(fixed_buf, elem_ty.offset, (seq_val as i8).to_le_bytes()),
+            PrimitiveType::U8 => write(fixed_buf, elem_ty.offset, (seq_val as u8).to_le_bytes()),
+            PrimitiveType::I16 => write(fixed_buf, elem_ty.offset, (seq_val as i16).to_le_bytes()),
+            PrimitiveType::U16 => write(fixed_buf, elem_ty.offset, (seq_val as u16).to_le_bytes()),
+            PrimitiveType::I32 => write(fixed_buf, elem_ty.offset, (seq_val as i32).to_le_bytes()),
+            PrimitiveType::U32 => write(fixed_buf, elem_ty.offset, (seq_val as u32).to_le_bytes()),
+            PrimitiveType::I64 => write(fixed_buf, elem_ty.offset, (seq_val as i64).to_le_bytes()),
+            PrimitiveType::U64 => write(fixed_buf, elem_ty.offset, (seq_val as u64).to_le_bytes()),
+            PrimitiveType::I128 => write(fixed_buf, elem_ty.offset, seq_val.to_le_bytes()),
+            PrimitiveType::U128 => write(fixed_buf, elem_ty.offset, (seq_val as u128).to_le_bytes()),
+            PrimitiveType::I256 => write(fixed_buf, elem_ty.offset, (i256::from(seq_val)).to_le_bytes()),
+            PrimitiveType::U256 => write(fixed_buf, elem_ty.offset, (u256::from(seq_val as u128)).to_le_bytes()),
+            PrimitiveType::Bool | PrimitiveType::F32 | PrimitiveType::F64 => {
+                panic!("`{:?}` is not a sequence integer type", &elem_ty.ty)
+            }
```

Comment:
It feels mildly odd that we panic here, but `unreachable_unchecked` above. I don't feel strongly, but I might replace this with `unreachable_unchecked` just for consistency's sake.